internal/commands/exec/exec.go

```diff
@@ -76,28 +81,53 @@ func RunExec(config ExecConfig) (ret int) {
 		}
 	}()
 
-	// First do the ACL Login, if necessary.
-	var consulClient *api.Client
-	var token string
-	var err error
-	if config.AuthConfig.Method != "" {
-		config.Logger.Trace("logging in to consul")
-		consulClient, token, err = login(config)
-		if err != nil {
-			config.Logger.Error("error logging into consul", "error", err)
-			return 1
-		}
-		config.Logger.Trace("consul login complete")
-	} else {
-		consulConfig := &config.ConsulConfig
-		consulConfig.Namespace = config.GatewayConfig.Namespace
-		consulClient, err = api.NewClient(consulConfig)
-		if err != nil {
-			config.Logger.Error("error reinitializing consul client", "error", err)
-			return 1
-		}
+	// ConsulServerConnMgr is the watcher for the Consul server addresses.
+	wctx, wcancel := context.WithTimeout(ctx, 3*time.Second)
+
+	// Release resources if consulServerConnMgr completes before timeout elapses
+	defer wcancel()
+
+	discoveryConfig := discovery.Config{
+		Addresses: config.ConsulHTTPAddress,
+		GRPCPort:  config.ConsulGRPCPort,
+		Credentials: discovery.Credentials{
+			Type: discovery.CredentialsTypeLogin,
+			Login: discovery.LoginCredential{
+				AuthMethod:  config.AuthConfig.Method,
+				Namespace:   config.AuthConfig.Method,
+				BearerToken: config.AuthConfig.Token,
+			},
+		},
+	}
+	if config.isTest {
+		discoveryConfig.ServerWatchDisabled = true
+	}
+
+	consulServerConnMgr, err := discovery.NewWatcher(wctx, discoveryConfig, config.Logger)
```

Comment:
So all of the discovery stuff should be moved into the new `consul.Client` interface implementation. You'll need to actually run a watcher goroutine to handle asynchronously swapping out the HTTP client when the returned server addresses change. I'll see if I can whip something up to demonstrate.

Reply:
@sarahalsmiller so here's actually pretty close to what I think you'll want, minus figuring out how to initialize all of this stuff 😅 https://github.com/hashicorp/consul-api-gateway/compare/sa-consul-server-connection-manager...consul-server-manager-demo